chaincode/src/fees/splitFeeBurnAndTransfer.ts

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) Gala Games Inc. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  BurnTokenQuantity,
+  FeeCodeSplitFormula,
+  PaymentRequiredError,
+  TokenClass,
+  TokenInstanceKey
+} from "@gala-chain/api";
+import BigNumber from "bignumber.js";
+
+import { burnTokens } from "../burns";
+import { transferToken } from "../transfer";
+import { GalaChainContext } from "../types";
+import { getObjectByKey } from "../utils";
+
+export interface splitFeeBurnAndTransferParams {
+  feeCode: string;
+  quantity: BigNumber;
+  galaCurrencyKey: TokenInstanceKey;
+  splitFormula: FeeCodeSplitFormula;
+}
+
+export async function splitFeeImmediatelyWithBurnAndTransfer(
+  ctx: GalaChainContext,
+  data: splitFeeBurnAndTransferParams
+): Promise<void> {
+  const payingUser = ctx.callingUser;
+  const { feeCode, galaCurrencyKey, splitFormula } = data;
+  const totalFeeQuantity = data.quantity;
+
+  const { collection, category, type, additionalKey } = galaCurrencyKey;
+  const galaCurrencyTokenClassKey = TokenClass.getCompositeKeyFromParts(TokenClass.INDEX_KEY, [
+    collection,
+    category,
+    type,
+    additionalKey
+  ]);
+  const galaCurrencyTokenClass = await getObjectByKey(ctx, TokenClass, galaCurrencyTokenClassKey);
+
+  const decimals: number = galaCurrencyTokenClass.decimals;
+
+  const [burnFeeQuantity, transferQuantities] = splitFormula.calculateAmounts(totalFeeQuantity, decimals);
+
+  const burnQuantity: BurnTokenQuantity = {
+    quantity: burnFeeQuantity,
+    tokenInstanceKey: galaCurrencyKey
+  };
+
+  await burnTokens(ctx, {
+    owner: payingUser,
+    toBurn: [burnQuantity]
+  }).catch((e) => {
+    throw new PaymentRequiredError(
+      `Failed to burnTokens for splitFeeFormula: ` +
+        `burnFeeQuantity: ${burnFeeQuantity.toString()}, ` +
+        `totalFeeQuantity: ${totalFeeQuantity.toString()}, with error: ${e.message}`,
+      { paymentQuantity: totalFeeQuantity.toString() }
+    );
+  });
+
+  for (const transfer of transferQuantities) {
+    await transferToken(ctx, {
+      from: payingUser,
+      to: transfer.transferToUser,
+      tokenInstanceKey: galaCurrencyKey,
+      quantity: transfer.transferQuantity,
+      allowancesToUse: [],
+      authorizedOnBehalf: undefined
+    }).catch((e) => {
+      throw new PaymentRequiredError(
+        `Failed to transferToken for splitFeeFormula: ` +
```

Comment:
do we throw that error in all cases? don't we want to check the error type?

Reply:
I agree with the sentiment - however it's a little complicated in this case. 

From a cursory look, I imagine the `ValidationFailedError` in `TokenBalance` with the string message `Insufficient Balance` would be a primary case to check for and throw `PaymentRequired`. 

We could assume other errors occur for reasons unrelated to the user being unable to pay, and maybe `PaymentRequired` is inappropriate. In those cases, maybe we would throw the provided error as-is rather than creating a new `PaymentRequiredError`.

However it could be brittle to code for this exact error - there are lots of other `ValidationFailedError`s thrown in `TokenBalance` and `TransferToken` and it's not clear to me what would be the best way to code for that specific error (e.g. `e.code === ErrorCodes.VALIDATION_FAILED && e.message === "Insufficient Balance" would break if someone edits the message in `TokenBalance.ts`). We'd probably want to encode a new, specific error to cover the "insufficient balance" case. But it's also possible other errors would additionally make sense to be included (like ObjectNotFound when a balance is not found, because the user has no balance whatsoever).

And it may be useful for consuming clients to get the `paymentQuantity` value regardless, for display to the end user or other uses. And this value is only provided if we throw `PaymentRequiredError` and include it. Given that 

1) whatever error happens here, was caused as a result of attempting to handle a required payment
2) clients might need or want the payment quantity 
3) the `PaymentRequiredError` is including the message from anything thrown - so something unrelated to an insufficient balance could still bubble up here

(another reason I'm hesitant to rework this too much is I'm currently in-process porting this to the SDK while still maintaining / updating new feature requests in the version in our internal repository, double-duty until we get this merged and incorporated back).

So I would argue yes, in this case we always throw a `PaymentRequiredError` for the reasons above. Or we pursue it as an upgrade later in a separate issue, after we get the migration settled. 

However, one useful change I could make is to include the error code/type and optional payload from the thrown error as part of the `PaymentRequiredError` - it currently propagates up the message but not the other diagnostics.